packages/mui-joy/src/Switch/Switch.tsx

```diff
@@ -0,0 +1,257 @@
+      cursor: 'default',
+      opacity: 0.6,
+    },
+    [`&.${switchClasses.focusVisible}`]: theme.focus.default,
+  };
+});
+
+const SwitchInput = styled('input', {
+  name: 'MuiSwitch',
+  slot: 'Input',
+  overridesResolver: (props, styles) => styles.input,
+})<{ ownerState: SwitchProps }>(() => ({
+  margin: 0,
+  height: '100%',
+  width: '100%',
+  opacity: 0,
+  position: 'absolute',
+  top: 0,
+  left: 0,
+  bottom: 0,
+  right: 0,
+  cursor: 'pointer',
+}));
+
+const SwitchTrack = styled('span', {
+  name: 'MuiSwitch',
+  slot: 'Track',
+  overridesResolver: (props, styles) => styles.track,
+})<{ ownerState: SwitchProps & { focusVisible: boolean } }>(() => ({
+  position: 'relative',
+  color: 'inherit',
+  height: 'var(--Switch-track-height)',
+  width: 'var(--Switch-track-width)',
+  display: 'block',
+  backgroundColor: 'currentColor',
+  borderRadius: 'var(--Switch-track-radius)',
+}));
+
+const SwitchThumb = styled('span', {
+  name: 'MuiSwitch',
+  slot: 'Thumb',
+  overridesResolver: (props, styles) => styles.thumb,
+})<{ ownerState: SwitchProps }>(() => ({
+  transition: 'left 0.2s',
+  position: 'absolute',
+  top: '50%',
+  left: 'calc(50% - var(--Switch-track-width) / 2 + var(--Switch-thumb-width) / 2 + var(--Switch-thumb-offset))',
+  transform: 'translate(-50%, -50%)',
+  width: 'var(--Switch-thumb-width)',
+  height: 'var(--Switch-thumb-size)',
+  borderRadius: 'var(--Switch-thumb-radius)',
+  backgroundColor: '#fff',
+  [`&.${switchClasses.checked}`]: {
+    left: 'calc(50% + var(--Switch-track-width) / 2 - var(--Switch-thumb-width) / 2 - var(--Switch-thumb-offset))',
+  },
+}));
+
+const Switch = React.forwardRef<HTMLSpanElement, SwitchProps>(function Switch(inProps, ref) {
+  const props = inProps;
+  const {
+    checked: checkedProp,
+    className,
+    component,
+    defaultChecked,
+    disabled: disabledProp,
+    onBlur,
+    onChange,
+    onFocus,
+    onFocusVisible,
+    readOnly: readOnlyProp,
+    required,
+    color = 'primary',
+    size,
+    ...otherProps
+  } = props;
+
+  const useSwitchProps = {
+    checked: checkedProp,
+    defaultChecked,
+    disabled: disabledProp,
+    onBlur,
+    onChange,
+    onFocus,
+    onFocusVisible,
+    readOnly: readOnlyProp,
+  };
+
+  const { getInputProps, checked, disabled, focusVisible, readOnly } = useSwitch(useSwitchProps);
+
+  const ownerState = {
+    ...props,
+    checked,
+    disabled,
+    focusVisible,
+    readOnly,
+    color,
+    size,
+  };
+
+  const classes = useUtilityClasses(ownerState);
+
+  return (
+    <SwitchRoot
+      ref={ref}
+      {...otherProps}
+      as={component}
+      ownerState={ownerState}
+      className={clsx(classes.root, className)}
+    >
+      <SwitchTrack ownerState={ownerState} className={clsx(classes.track)} />
+      <SwitchThumb ownerState={ownerState} className={clsx(classes.thumb)} />
+      <SwitchInput ownerState={ownerState} {...getInputProps()} className={clsx(classes.input)} />
```

Comment:
I see there's no ability to customize the props of inner components (`componentsProps` is not exposed). Is it by design or do you plan to still add it?

Reply:
Good catch, I will add it.